File: Introduction to Algorithms/Integer List ADT in C/List.c
//Program 2: Doubly Linked List
//Justin Wong jujwong@ucsc.edu
//CS101 Tantalo Fall 14


#include <stdlib.h>
#include <assert.h>
#include <stdio.h>
#include "List.h"



typedef struct nodeObj* node;

struct nodeObj {
  int data;
  node prev;
  node next;
};

typedef struct ListObj {
  node curr;
  node front;
  node back;
  int length;
  int index;
} ListObj;

// constructor for a new list
List newList() {
  List new = malloc (sizeof(struct ListObj));
  new->curr = NULL;
  new->front = NULL;
  new->back = NULL;
  new->length = 0;
  new->index = -1;
  return new;
}

//Frees list, then frees the pointer itself
void freeList (List* pL){
  if ((*pL) == NULL) {
    printf("Error: calling freeList() on NULL list reference");
    return;
  }
  clear(*pL);
  free(*pL);
  *pL = NULL;
}

// Returns number of elements in this list.
int length(List L){
  if (L == NULL){
    printf("Error: calling length() on NULL list reference");
    return -1;
  } 
  return L->length;
}

 // Returns the index of the cursor element in this list, or 
 // returns -1 if the cursor element is undefined.
int getIndex(List L){
  if (L == NULL){
    printf("Error: calling getIndex() on NULL list reference");
    return -1;
  }
  return L->index;
}

// Returns front element in this list. 
//Pre: length()>0
int front(List L){
  if (L == NULL){
    printf("Error: calling front() on NULL list reference");
    return -1;
  }
  if (L->length > 0) return L->front->data;
  return -1;
}

// Returns back element in this List. 
//Pre: length()>0
int back(List L){
  if (L == NULL){
    printf("Error: calling back() on NULL list reference");
    return -1;
  }
  if (L->length > 0) return L->back->data;
  return -1;
}

// Returns curr element in the list 
// Pre: length()>0, getIndex()>=0
int getElement(List L){
  if (L == NULL){
    printf("Error: calling getElement() on NULL list reference");
    return -1;
  }
  if (L->length > 0 && L->index >=0){
    return L->curr->data;
  }
  return -1;
}

//Returns true if the two lists are the same
//checks the indices of two lists
//and also checks that they are same length
int equals (List A, List B){
  if (A == NULL || B == NULL){
    printf("Error: calling equals() on NULL list reference");
    return -1;
  }
  node right = A->front;
  node left = B->front;
  while (right != NULL && left !=NULL){
    if(right->data != left->data) return 0;
    right = right->next;
    left = left->next;
  }
  if(right != NULL || left != NULL) return 0;
  return 1;
}

// Re-sets the list to the empty state
void clear(List L){
  if (L == NULL){
    printf("Error: calling clear() on NULL list reference");
    return;
  }
  node front = L->front;
  while (front != NULL){
    node temp = front->next;
    free(front);
    front = temp;
  }
  L->front = NULL;
  L->back = NULL;
  L->curr = NULL;
  L->length = 0;
  L->index = -1;
}

// If 0<=i<=length()-1, moves the cursor to the element 
// at index i, otherwise the cursor becomes undefined.
void moveTo(List L, int i){
  if (L == NULL){
    printf("Error: calling moveTo() on NULL list reference");
    return;
  }
  L->curr = L->front;
  L->index = i;
  if(i >= 0 && i < L->length){
    for(; i > 0; i--){
      L->curr = L->curr->next;
    }
  } else {
    L->curr = NULL;
    L->index = -1;
  }
}

// If 0<getIndex()<=length()-1, moves the cursor one step toward the
// front of the list. If getIndex()==0, cursor becomes undefined. 
// If getIndex()==-1, cursor remains undefined. This operation is
// equivalent to moveTo(getIndex()-1).
void movePrev(List L){
  if (L == NULL) {
    printf("Error: calling movePrev() on NULL list reference");
    return;
  }
  if(L->index > 0 && L->index < L->length){
    L->curr = L->curr->prev;
    L->index--;
  } else {
    L->curr = NULL;
    L->index = -1;
  }
}

// If 0<=getIndex()<length()-1, moves the cursor one step toward the
// back of the list. If getIndex()==length()-1, cursor becomes
// undefined. If index==-1, cursor remains undefined. This
// operation is equivalent to moveTo(getIndex()+1).
void moveNext(List L){
  if (L == NULL) {
    printf("Error: calling moveNext() on NULL list reference");
    return;
  }
  if(L->index >= 0 && L->index < L->length - 1){
    L->curr = L->curr->next;
    L->index++;
  } else {
    L->curr = NULL;
    L->index = -1;
  }
}

// Inserts new element before front element in this List
// makes new node then checks if node is only element in list
// to determine the front, and back
void prepend(List L, int data){
  if(L == NULL) {
    printf("Error: calling prepend() on NULL list reference");
    return;
  }
  L->length++;
  node node = malloc(sizeof(struct nodeObj));
  node->data = data;
  node->prev = NULL;
  node->next = L->front;
  if (L->front == NULL){
    L->back = node;
  } else {
    L->front->prev = node;
  }  
  L->front = node;
  if (L->index >= 0) L->index++;
}

// Inserts new element after back element in this List.
// makes new node and checks if node is only element in list
// to determine the front, and back
void append(List L, int data){
  if(L == NULL){
    printf("Error: calling append() on NULL list reference");
    return;
  }
  L->length++;
  node node = malloc(sizeof(struct nodeObj));
  node->data = data;
  node->prev = L->back;
  node->next = NULL;
  if(L->back == NULL){
    L->front = node;
  } else {
    L->back->next = node;
  }
  L->back = node;
}

// Inserts new element before cursor element
// Pre: length()>0, getIndex()>=0
void insertBefore(List L, int data){
  if (L == NULL){
    printf("Error: calling insertBefore() on NULL list reference");
    return;
  }
  if (!(L->length > 0 && L->index >= 0)) return;
  L->length++;
  node node = malloc(sizeof(struct nodeObj));
  node->data = data;
  node->prev = L->curr->prev;
  node->next = L->curr;
  if(node->prev != NULL){
    L->curr->prev->next = node; 
  } else {
    L->front = node;
  }
  L->curr->prev = node;
  L->index++;
}

// Inserts new element before cursor element
//Pre: length()>0, getIndex()>=0
void insertAfter(List L, int data){
  if(L == NULL){
    printf("Error: calling insertAfter() on NULL list reference");
    return;
  }
  if (!(L->length > 0 && L->index >= 0)) return;
  L->length++;
  node node = malloc(sizeof(struct nodeObj));
  node->data = data;
  node->prev = L->curr;
  node->next = L->curr->next; 
  if(node->next != NULL){
    L->curr->next->prev = node;
  } else {
    L->back = node;
  }
  L->curr->next = node;
}

// Deletes the front element in this List. 
//Pre: length()>0
void deleteFront(List L){
  if(L == NULL){
    printf("Error: calling deleteFront() on NULL list reference");
    return;
  }
  if(L->length > 0){
    L->length--;
    if(L->index == 0) L->curr = NULL;
    if(L->back == L->front) L->back = NULL;
    node temp = L->front;
    L->front = L->front->next;
    free(temp);
    L->front->prev = NULL;
    if(L->index != -1) L->index--;
  }
}

// Deletes the back element in this List. 
//Pre: length()>0
void deleteBack(List L){
  if(L == NULL){
    printf("Error: calling deleteBack() on NULL list reference");
    return;
  }
  if(L->length > 0){
    L->length--;
    if(L->index == L->length - 1){
      L->curr = NULL;
      L->index = -1;
    }
    if(L->back == L->front) L->front = NULL;
    node temp = L->back;
    L->back = L->back->prev;
    free(temp);
    L->back->next = NULL;
  }
}

// Deletes cursor element in this List. 
// Cursor is undefined after this operation. 
// Pre: length()>0, getIndex()>=0
void delete(List L){
  if(L == NULL){
    printf("Error: calling delete() on NULL list reference");
    return;
  }
  if(L->curr == L->front){
    deleteFront(L);
  } else if (L->curr == L->back){
    deleteBack(L);
  }
  if (L->length > 0 && L->index > 0){
    L->length--;
    node temp = L->curr;
    L->curr->next->prev = L->curr->prev;
    L->curr->prev->next = L->curr->next;
    L->curr = NULL;
    L->index = -1;
    free(temp);
  }
}

//Prints list
void printList (FILE *out, List L){
  if (L == NULL){
    printf("Error: calling printList() on NULL list reference");
    return;
  }
  if (out == NULL){
    printf("Error: out file is NULL");
    return;
  }
  node temp = L->front;
  while (temp != NULL){
    fprintf(out, "%d ", temp->data);
    temp = temp->next;
  }
}

// Returns a new list representing the same integer sequence as this
// list. The cursor in the new list is undefined, regardless of the
// state of the cursor in this List. This List is unchanged
List copyList(List L){
  if(L == NULL){
    printf("Error: calling copyList() on NULL list reference");
    return NULL;
  }
  List copy = newList();
  for(node temp = L->front; temp != NULL; temp = temp->next){
    append(copy, temp->data);
  }
  return copy;
}

// Returns a new List which is the concatenation of the two
List concat(List A, List B){
  if(A == NULL || B == NULL) printf("Error: List is NULL");
  List catList = copyList(A);
  for(node temp = B->front; temp != NULL; temp = temp->next){
    append(catList, temp->data);
  }
  return catList;
}

File: Introduction to Algorithms/Integer List ADT in Java/List.java
// Program 1: Doubly Linked List
//Justin Wong jujwong

import java.util.*;
import java.util.NoSuchElementException;

class List{

   //defining the node
   private class Node{
      //fields
      int data;
	  Node next;
	  Node prev;
	  
	  //constructor
	  Node(int data){
	     this.data = data;
		 next = prev = null;
	  }
	  public String toString() {return String.valueOf(data);}
   }
   
   //Fields
   private Node front;
   private Node back;
   private Node curr;
   private int length;
   
   //constructor
    List() {
	front = back = curr = null; 
	length = 0;
	}
	
	//Access Functions
   boolean isEmpty() {return length == 0;}
   boolean off() {return (this.curr == null);}
   
   int length() {return length;}
   
   //Pre: length() > 0
   int front() {
      if (this.isEmpty()){
      throw new RuntimeException("Error: front called on empty list");}
	  return front.data;
   }
   
   //Pre: length() > 0
   int back() {
      if (this.isEmpty()){
      throw new RuntimeException("Error: back called on empty list");}
	  return back.data;
   }
   
   //returns the index of cursor element in list or -1 if cursor is undefined
   int getIndex(){
      int index;
	  if(this.off()){
	     index = -1;
		 return index;
	  }else{
      Node temp = new Node(this.front.data);
	  temp = this.front;
      for(index = 0; this.curr != temp; ++index){
	     temp = temp.next;
	  }
	  return index;
	  }
   }
   
   // Returns cursor element in this list. 
   // Pre: length()>0, getIndex()>=0
   int getElement(){
   if(this.isEmpty()){
      throw new RuntimeException("Error: empty list");
   }else if(this.getIndex() < 0){
	  throw new RuntimeException("Error: out of bounds");
   }
   return this.curr.data;
   }
   
   //returns true if this List and L are same integer sequence, cursor is ignored in both lists
   boolean equals(List L){
      boolean flag = true;
	  Node M = this.front;
	  Node N = L.front;
	  if(this.length == L.length){
	     while( flag && M != null){
		    flag = (N.data == M.data);
			N = N.next;
			M = M.next;
		 }
		 return flag;
	  } else {
	     return false;
	  }
   }
   
   //Manipulation Procedures
   
   //resets list to empty state
   void clear(){
      this.back = null;
	  while ( !this.isEmpty()){
	     this.curr = this.front;
		 this.curr = this.curr.next;
		 if ( this.curr != null) this.curr.prev = null;
		 this.front.next = null;
		 --length;
	  }
	  this.front = null;
   }
   
   //If 0<=i<=length()-1, moves the cursor to the element 
   //at index i, otherwise the cursor becomes undefined.
   void moveTo(int i){
      if(i < 0 && i > (this.length)){
	     throw new RuntimeException("Error: not in list");
	   }
       this.curr = this.front;
       for(int t = 0; t < i; ++t){
          this.curr = this.curr.next;
       }	   
	  }
   
   // If 0<getIndex()<=length()-1, moves the cursor one step toward the
   // front of the list. If getIndex()==0, cursor becomes undefined. 
   // If getIndex()==-1, cursor remains undefined. This operation is
   // equivalent to moveTo(getIndex()-1).
   void movePrev(){
      if(this.isEmpty()){
	     throw new RuntimeException("Error: empty list");
	  }else if(this.getIndex() == -1){
	     this.curr = null;
	     //throw new RuntimeException("Error: pointer off list");
	  }else if(this.getIndex() == 0){
	     this.curr = null;
	  }
      else if((this.getIndex() > 0) && (this.getIndex() <=(this.length() - 1))){
	     //this.curr = this.curr.prev; //this line doesnt work
		 this.moveTo(this.getIndex() -1); //works
		 }
   }
   
   
   // If 0<=getIndex()<length()-1, moves the cursor one step toward the
   // back of the list. If getIndex()==length()-1, cursor becomes
   // undefined. If index==-1, cursor remains undefined. This
   // operation is equivalent to moveTo(getIndex()+1).
   void moveNext(){
      if(this.isEmpty()){
	     throw new RuntimeException("Error: empty list");
	  }else if(this.off()){
	     this.curr = null;
	  }else if (this.getIndex() == (this.length() -1)){
	     this.curr = null;
	  }
      if((this.getIndex() >= 0) && ((this.length() - 1) > this.getIndex())){
	     this.curr = this.curr.next;
		 }
   }
   
   // Inserts new element before front element in this List.
   void prepend(int data){
      Node temp = new Node(data);
	  if (this.isEmpty()){front = back = temp;}
	  else { temp.next = front; temp.prev = temp; front = temp;}
	  ++length;
   }
   
   // Inserts new element after back element in this List.
   void append(int data){
      Node temp = new Node(data);
	  if(this.isEmpty()){front = back = temp;}
	  else { temp.prev = back; back.next = temp; back = temp;}
	  ++length;
   }
   
   // Inserts new element before cursor element in this
   // List. Pre: length()>0, getIndex()>=0
   void insertBefore(int data){
      if(this.length < 0){
        throw new RuntimeException("Error: empty list");
        }
      else if(this.getIndex() < 0){
         throw new RuntimeException("Error: cursor is undefined");
      }
      Node temp = new Node(data);
	  if(curr.prev == null){
	     temp.next = front;
		 front.prev = temp;
		 front = temp;
	  } else {
	     temp.prev = curr.prev;
		 temp.next = curr;
		 curr.prev.next = temp;
		 curr.prev = temp;
	  }
	  ++length;
   }
   
   // Inserts new element after cursor element in this 
   // List. Pre: length()>0, getIndex()>=0
   void insertAfter(int data){
      if(this.length < 0){
	     throw new RuntimeException("Error: length is < 0");
	  }
	  else if(this.getIndex() < 0){
	     throw new RuntimeException("Error: cursor is undefined");
	  }
	  Node temp = new Node(data);
	  if(curr.next == null){
	     back.next = temp;
		 back = temp;
		 back.prev = curr;
	  }else{
	     temp.prev = curr;
		 temp.next = curr.next;
		 curr.next.prev = temp;
		 curr.next = temp;
	  }
	  ++length;
   }
   
   // Deletes the front element in this List. Pre: length()>0
   // doesn't return a node
   void deleteFront(){
      if(this.isEmpty()){
	     throw new RuntimeException("Error: List is empty");
	  }
	  if(this.length > 1){ //if there is more than one element in list
	     front = front.next;
	     front.prev = null;
	  }else{ //if there is only 1 element in list
	  front = back = curr = null;
	  }
	  --length;
   }
   
   // Deletes the back element in this List. Pre: length()>0
   void deleteBack(){
      if(this.isEmpty()){
	     throw new RuntimeException("Error: List is empty");
	  }
	  if(this.length > 1){
	     back = back.prev;
		 back.next = null;
	  }else{
	     front = back = curr = null;
	  }
	  --length;
   }
   
   
   // Deletes cursor element in this List. Cursor is undefined after this 
  // operation. Pre: length()>0, getIndex()>=0
   void delete(){
      if(this.length < 0){
	     throw new RuntimeException("Error: length is < 0");
	  }
	  else if(this.getIndex() < 0){
	     throw new RuntimeException("Error: cursor is undefined");
	  }
	  if(this.length > 1){
	     if(curr.prev == null){
		    front = curr.next;
			curr.next = null;
		 }else if(curr.next == null){
		    back = curr.prev;
			curr.prev = null;
		 }else{
		    curr.prev.next = curr.next;
			curr.next.prev = curr.prev;
			curr.prev = curr.next = null;
		 }
		 curr = null;
	  }
	  else{ front = back = curr = null;}
	  --length;
   }
   
   //Other methods
   
   public String toString(){
      String str = "";
	  for (Node temp = front; temp != null; temp = temp.next){
	     str += temp.toString() + " ";
	  }
	  return str;
   }
   
   // Returns a new list representing the same integer sequence as this
   // list. The cursor in the new list is undefined, regardless of the
   // state of the cursor in this List. This List is unchanged.
   List copy(){
      List copyL = new List();
	  Node temp = this.front;
	  while(temp != null){
	     copyL.append(temp.data);
		 temp = temp.next;
	  }
      return copyL;
   }
   
    // Returns a new List which is the concatenation of 
    // this list followed by L. The cursor in the new list
    // is undefined, regardless of the states of the cursors
    // in this list and L. The states of this list and L are
    // unchanged.
   List concat(List L){
      List list = new List();
	  List catList = new List();
	  list = this.copy();
	  catList = L.copy();
	  list.back.next = catList.front;
	  catList.front.prev = list.back;
	  list.back = catList.back;
	  list.length += L.length;
	  return list;
   }
   
   
   
   
   
}
